File: testutils/pn532/pn532.py
#  Copyright (C) 2024 The Android Open Source Project
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

# Lint as: python3

import logging
import serial
from . import tag
from binascii import hexlify
from serial.tools.list_ports import comports
from mobly import logger as mobly_logger

GET_FIRMWARE_VERSION = 0x02
SAM_CONFIGURATION = 0x14
RF_CONFIGURATION = 0x32
IN_DATA_EXCHANGE = 0x40
IN_COMMUNICATE_THRU = 0x42
IN_LIST_PASSIVE_TARGET = 0x4A
WRITE_REGISTER = 0x08
LONG_PREAMBLE = bytearray(20)
TG_INIT_AS_TARGET = 0x8C

def crc16a(data):
    w_crc = 0x6363
    for byte in data:
        byte = byte ^ (w_crc & 0x00FF)
        byte = (byte ^ (byte << 4)) & 0xFF
        w_crc = ((w_crc >> 8) ^ (byte << 8) ^ (byte << 3) ^ (byte >> 4)) & 0xFFFF
    return bytes([w_crc & 0xFF, (w_crc >> 8) & 0xFF])


def with_crc16a(data):
    return bytes(data) + crc16a(data)


class PN532:

    def __init__(self, path):
        """Initializes device on path, or first available serial port if none is provided."""
        if len(comports()) == 0:
            raise IndexError(
                "Could not find device on serial port, make sure reader is plugged in."
            )
        if len(path) == 0:
            path = comports()[0].device
        self.log = mobly_logger.PrefixLoggerAdapter(
            logging.getLogger(),
            {
                mobly_logger.PrefixLoggerAdapter.EXTRA_KEY_LOG_PREFIX: (
                    f"[PN532|{path}]"
                )
            },
        )
        self.log.debug("Serial port: %s", path)
        self.device = serial.Serial(path, 115200, timeout=0.5)

        self.device.flush()
        self.device.write(LONG_PREAMBLE + bytearray.fromhex("0000ff00ff00"))
        self.device.flushInput()
        if not self.verify_firmware_version():
            raise RuntimeError("Could not verify PN532 firmware on serial path " + path)
        rsp = self.send_frame(
            LONG_PREAMBLE + self.construct_frame([SAM_CONFIGURATION, 0x01, 0x00]),
            1,
            )
        if not rsp:
            raise RuntimeError("No response for SAM configuration.")

        # Disable retries
        self.device.flushInput()
        rsp = self.send_frame(
            self.construct_frame(
                [
                    RF_CONFIGURATION,
                    0x05,
                    0x00,  # MxRtyATR
                    0x00,  # MxRtyPSL
                    0x00,  # MxRtyPassiveActivation
                ]
            ),
            1,
        )
        if not rsp:
            raise RuntimeError("No response for RF configuration.")

    def verify_firmware_version(self):
        """Verifies we are talking to a PN532."""
        self.log.debug("Checking firmware version")
        rsp = self.send_frame(
            LONG_PREAMBLE + self.construct_frame([GET_FIRMWARE_VERSION])
        )

        if not rsp:
            raise RuntimeError("No response for GetFirmwareVersion")

        if rsp[0] != GET_FIRMWARE_VERSION + 1 or len(rsp) != 5:
            self.log.error("Got unexpected response for GetFirmwareVersion")
            return False

        return rsp[1] == 0x32

    def poll_a(self):
        """Attempts to detect target for NFC type A."""
        self.log.debug("Polling A")
        rsp = self.send_frame(
            self.construct_frame([IN_LIST_PASSIVE_TARGET, 0x01, 0x00])
        )
        if not rsp:
            raise RuntimeError("No response for send poll_a frame.")

        if rsp[0] != IN_LIST_PASSIVE_TARGET + 1:
            self.log.error("Got unexpected command code in response")
        del rsp[0]

        num_targets = rsp[0]
        if num_targets == 0:
            return None
        del rsp[0]

        target_id = rsp[0]
        del rsp[0]

        sense_res = rsp[0:2]
        del rsp[0:2]

        sel_res = rsp[0]
        self.log.debug("Got tag, SEL_RES is %02x", sel_res)
        del rsp[0]

        nfcid_len = rsp[0]
        del rsp[0]
        nfcid = rsp[0:nfcid_len]
        del rsp[0:nfcid_len]

        ats_len = rsp[0]
        del rsp[0]
        ats = rsp[0 : ats_len - 1]
        del rsp[0 : ats_len - 1]

        return tag.TypeATag(self, target_id, sense_res, sel_res, nfcid, ats)

    def initialize_target_mode(self):
        """Configures the PN532 as target."""
        self.log.debug("Initializing target mode")
        self.send_frame(
            self.construct_frame([TG_INIT_AS_TARGET,
                                  0x05, #Mode
                                  0x04, #SENS_RES (2 bytes)
                                  0x00,
                                  0x12, #nfcid1T (3 BYTES)
                                  0x34,
                                  0x56,
                                  0x20, #SEL_RES
                                  0x00, #FeliCAParams[] (18 bytes)
                                  0x00,
                                  0x00,
                                  0x00,
                                  0x00,
                                  0x00,
                                  0x00,
                                  0x00,
                                  0x00,
                                  0x00,
                                  0x00,
                                  0x00,
                                  0x00,
                                  0x00,
                                  0x00,
                                  0x00,
                                  0x00,
                                  0x00,
                                  0x00,#NFCID3T[] (10 bytes)
                                  0x00,
                                  0x00,
                                  0x00,
                                  0x00,
                                  0x00,
                                  0x00,
                                  0x00,
                                  0x00,
                                  0x00,
                                  0x00, #LEN Gt
                                  0x00, #LEN Tk
                                  ]))

    def poll_b(self):
        """Attempts to detect target for NFC type B."""
        self.log.debug("Polling B")
        rsp = self.send_frame(
            self.construct_frame([IN_LIST_PASSIVE_TARGET, 0x01, 0x03, 0x00])
        )
        if not rsp:
            raise RuntimeError("No response for send poll_b frame.")

        if rsp[0] != IN_LIST_PASSIVE_TARGET + 1:
            self.log.error("Got unexpected command code in response")
        del rsp[0]

        afi = rsp[0]

        deselect_command = 0xC2
        self.send_broadcast(bytearray(deselect_command))

        wupb_command = [0x05, afi, 0x08]
        self.send_frame(
            self.construct_frame([WRITE_REGISTER, 0x63, 0x3D, 0x00])
        )
        rsp = self.send_frame(
            self.construct_frame(
                [IN_COMMUNICATE_THRU] + list(with_crc16a(wupb_command))
            )
        )
        if not rsp:
            raise RuntimeError("No response for WUPB command")

        return tag.TypeBTag(self, 0x03, rsp)

    def send_broadcast(self, broadcast):
        """Emits broadcast frame with CRC. This should be called after poll_a()."""
        self.log.debug("Sending broadcast %s", hexlify(broadcast).decode())

        # Adjust bit framing so all bytes are transmitted
        self.send_frame(self.construct_frame([WRITE_REGISTER, 0x63, 0x3D, 0x00]))
        rsp = self.send_frame(
            self.construct_frame([IN_COMMUNICATE_THRU] + list(with_crc16a(broadcast)))
        )
        if not rsp:
            raise RuntimeError("No response for send broadcast.")

    def transceive(self, data):
        """Sends data to device and returns response."""
        self.log.debug("Transceive")
        rsp = self.send_frame(self.construct_frame([IN_DATA_EXCHANGE] + list(data)), 5)

        if not rsp:
            return None

        if rsp[0] != IN_DATA_EXCHANGE + 1:
            self.log.error("Got unexpected command code in response")
        del rsp[0]

        if rsp[0] != 0:
            self.log.error("Got error exchanging data")
            return None
        del rsp[0]

        return rsp

    def mute(self):
        """Turns off device's RF antenna."""
        self.log.debug("Muting")
        self.send_frame(self.construct_frame([RF_CONFIGURATION, 0x01, 0x02]))

    def construct_frame(self, data):
        """Construct a data fram to be sent to the PN532."""
        # Preamble, start code, length, length checksum, TFI
        frame = [
            0x00,
            0x00,
            0xFF,
            (len(data) + 1) & 0xFF,
            ((~(len(data) + 1) & 0xFF) + 0x01) & 0xFF,
            0xD4,
            ]
        data_sum = 0xD4

        # Add data to frame
        for b in data:
            data_sum += b
            frame.append(b)
        frame.append(((~data_sum & 0xFF) + 0x01) & 0xFF)  # Data checksum

        frame.append(0x00)  # Postamble
        self.log.debug("Constructed frame " + hexlify(bytearray(frame)).decode())

        return bytearray(frame)

    def send_frame(self, frame, timeout=0.5):
        """
        Writes a frame to the device and returns the response.
        """
        self.device.write(frame)
        return self.get_device_response(timeout)

    def reset_buffers(self):
        self.device.reset_input_buffer()
        self.device.reset_output_buffer()

    def get_device_response(self, timeout=0.5):
        """
        Confirms we get an ACK frame from device, reads response frame, and writes ACK.
        """
        self.device.timeout = timeout
        frame = bytearray(self.device.read(6))

        if (len(frame)) == 0:
            self.log.error("Did not get response from PN532")
            return None

        if hexlify(frame).decode() != "0000ff00ff00":
            self.log.error("Did not get ACK frame, got %s", hexlify(frame).decode())

        frame = bytearray(self.device.read(6))

        if (len(frame)) == 0:
            return None

        if hexlify(frame[0:3]).decode() != "0000ff":
            self.log.error(
                "Unexpected start to frame, got %s", hexlify(frame[0:3]).decode()
            )

        data_len = frame[3]
        length_checksum = frame[4]
        if (length_checksum + data_len) & 0xFF != 0:
            self.log.error("Frame failed length checksum")
            return None

        tfi = frame[5]
        if tfi != 0xD5:
            self.log.error(
                "Unexpected TFI byte when performing read, got %02x", frame[5]
            )
            return None

        data_packet = bytearray(
            self.device.read(data_len - 1)
        )  # subtract one since length includes TFI byte.
        data_checksum = bytearray(self.device.read(1))[0]
        if (tfi + sum(data_packet) + data_checksum) & 0xFF != 0:
            self.log.error("Frame failed data checksum")

        postamble = bytearray(self.device.read(1))[0]
        if postamble != 0x00:
            if tfi != 0xD5:
                self.log.error(
                    "Unexpected postamble byte when performing read, got %02x", frame[4]
                )

        self.device.timeout = 0.5
        self.device.write(
            bytearray.fromhex("0000ff00ff00")
        )  # send ACK frame, there is no response.

        self.log.debug(
            "Received frame %s%s",
            hexlify(frame).decode(),
            hexlify(data_packet).decode(),
        )

        return data_packet

File: testutils/pn532/src/com/android/nfc/pn532/PN532.kt
/*
 * Copyright (C) 2024 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.nfc.pn532

import android.hardware.usb.UsbDevice
import android.hardware.usb.UsbDeviceConnection
import android.util.Log
import com.google.errorprone.annotations.CanIgnoreReturnValue

/**
 * Handles communication with PN532 given a UsbDevice and UsbDeviceConnection object. Relevant
 * protocols of device are located at https://www.nxp.com/docs/en/user-guide/141520.pdf
 */
class PN532(val device: UsbDevice, val connection: UsbDeviceConnection) {
    private val transportLayer: TransportLayer

    init {
        Log.d(TAG, "Starting initialization")
        transportLayer = TransportLayer(device, connection)

        // Wake up device and send initial configs
        val rsp = transportLayer.write(LONG_PREAMBLE + ACK)
        if (!rsp) {
            Log.e(TAG, "Got error while waking device up.")
        }
        val samRsp = sendFrame(constructFrame(byteArrayOf(SAM_CONFIGURATION, 0x01, 0x00)))
        if (samRsp == null) {
            Log.e(TAG, "Did not get a response after sending SAM config command")
        }
        val rfRsp = sendFrame(constructFrame(byteArrayOf(RF_CONFIGURATION, 0x05, 0x01, 0x00, 0x01)))
        if (rfRsp == null) {
            Log.e(TAG, "Did not get a response after sending RF config command")
        }
    }

    /** Polls for NFC Type-A. Returns tag if discovered. */
    fun pollA(): TypeATag? {
        Log.d(TAG, "Polling A")
        val rsp =
            sendFrame(constructFrame(byteArrayOf(IN_LIST_PASSIVE_TARGET, 0X01, 0X00)), timeout = 5000)
        if (rsp == null || rsp.size == 0) return null
        if (rsp[0] != (IN_LIST_PASSIVE_TARGET + 1).toByte()) {
            Log.e(TAG, "Got unexpected command code in response")
            return null
        }
        val targetData = rsp.drop(1).toByteArray()
        if (targetData.size < 6 || targetData[0] == 0.toByte()) {
            Log.d(TAG, "No tag found")
            return null
        }
        Log.d(TAG, "Tag found. Response: " + rsp.toHex())

        // Page 116 of https://www.nxp.com/docs/en/user-guide/141520.pdf for response format info
        val targetId = targetData[1]
        val senseRes = targetData.slice(2..3).toByteArray()
        val selRes = targetData[4]
        val nfcIdLen = targetData[5]

        if (6 + nfcIdLen > targetData.size) {
            Log.e(
                TAG,
                "Corrupt data - expected remaining size of list to be at least: " +
                        nfcIdLen +
                        " , but was " +
                        (targetData.size - 6),
            )
            return null
        }

        val nfcIdList = targetData.slice(6..6 + nfcIdLen - 1).toByteArray()
        val ats = targetData.drop(6 + nfcIdLen).toByteArray()
        if (ats.size == 0) {
            Log.e(TAG, "Corrupt data - expected ATS information")
            return null
        }
        val atsLen = ats[0]
        val atsList = ats.drop(1).toByteArray()
        if (atsList.size != atsLen.toInt() - 1) {
            Log.e(
                TAG,
                "Corrupt data - expected list of size " +
                        (atsLen.toInt() - 1) +
                        " , but was " +
                        (atsList.size),
            )
            return null
        }

        return TypeATag(this, targetId, senseRes, selRes, nfcIdList, atsList)
    }

    /** Polls for NFC Type-B */
    fun pollB() {
        Log.d(TAG, "Polling B")
        val rsp = sendFrame(constructFrame(byteArrayOf(IN_LIST_PASSIVE_TARGET, 0x01, 0x03, 0x00)))
        if (rsp == null) {
            Log.e(TAG, "Did not get a  response after sending polling command.")
        }
    }

    /** Emits broadcast frame with CRC. Call this after pollA() to send a custom frame */
    fun sendBroadcast(broadcast: ByteArray) {
        Log.d(TAG, "sendBroadcast: " + broadcast.toHex())
        val writeRsp = sendFrame(constructFrame(byteArrayOf(WRITE_REGISTER, 0X63, 0X3D, 0X00)))
        if (writeRsp == null) {
            Log.e(TAG, "Did not get valid response after sending WRITE_REGISTER command")
        }
        val broadcastRsp = sendFrame(constructFrame(byteArrayOf(IN_COMMUNICATE_THRU) +
                withCrc16a(broadcast)))
        if (broadcastRsp == null) {
            Log.e(TAG, "Did not get valid response after sending broadcast")
        }
    }

    /** Send command to PN-532 and receive response. */
    fun transceive(data: ByteArray): ByteArray? {
        Log.d(TAG, "Transceiving: " + data.toHex())
        val response = sendFrame(constructFrame(byteArrayOf(IN_DATA_EXCHANGE) + data))
        if (response == null) return null
        Log.d(TAG, "Response: " + response.toHex())

        if (response[0] != (IN_DATA_EXCHANGE + 1).toByte()) {
            Log.e(TAG, "Got unexpected command code in response")
        }
        if (response[1] != 0.toByte()) {
            Log.e(TAG, "Got error exchanging data")
            return null
        }
        return response.drop(2).toByteArray()
    }

    /** Mute reader. Should be called after each polling loop. */
    fun mute() {
        Log.d(TAG, "Muting PN532")
        val rsp = sendFrame(constructFrame(byteArrayOf(RF_CONFIGURATION, 0x01, 0x02)))
        if (rsp == null) {
            Log.e(TAG, "Did not get valid response after muting.")
        }
    }

    @CanIgnoreReturnValue
    private fun sendFrame(frame: ByteArray, timeout: Long = 500.toLong()): ByteArray? {
        transportLayer.write(frame)
        return getDeviceResponse(timeout)
    }

    private fun isAckFrame(frame: ByteArray): Boolean {
        return frame.toHex().contentEquals(ACK.toHex())
    }

    private fun getDeviceResponse(timeoutMs: Long = 500.toLong()): ByteArray? {
        // First response from device should be ACK frame
        val data = transportLayer.read(timeoutMs, numBytes = MAX_READ_BUFFER_SIZE)
        if (data == null || data.size < 6) return null

        val firstFrame = data.slice(0..5).toByteArray()
        if (!isAckFrame(firstFrame)) {
            Log.w(TAG, "Did not get ack frame - got " + firstFrame.toHex())
            return null
        } else {
            Log.d(TAG, "Got ack frame")
        }

        // Response will either be appended to first read, or will require an additional read
        var responseFrame: ByteArray? = data.drop(6).toByteArray()

        // Some instances require a second read of data
        var secondRead = false
        if (responseFrame?.size == 0) {
            responseFrame = transportLayer.read(timeoutMs, numBytes = MAX_READ_BUFFER_SIZE)
            secondRead = true
        }
        if (responseFrame == null || responseFrame.size == 0) {
            Log.d(TAG, "No additional data")
            return null
        }
        if (responseFrame.size < 6) {
            Log.w(TAG, "Expected at least 6 bytes of response data. Got " + responseFrame.size)
            return null
        }

        if (isAckFrame(responseFrame)) {
            Log.d(TAG, "Got another ack frame")
            return null
        }

        if (!responseFrame.slice(0..2).toByteArray().toHex().contentEquals("0000ff")) {
            Log.e(
                TAG,
                "Unexpected start to frame - got " +
                        responseFrame.slice(0..2).toByteArray().toHex() +
                        ", expected " +
                        "0000ff",
            )
        } else {
            Log.d(TAG, "Correct start to frame")
        }

        val isExtendedFrame = responseFrame.slice(3..4).toByteArray().toHex().contentEquals("ffff")
        val expectedDataLength: UInt
        if (isExtendedFrame) {
            if (responseFrame.size < EXTENDED_FRAME_MIN_SIZE) {
                Log.e(
                TAG,
                "Expected at least " +
                    EXTENDED_FRAME_MIN_SIZE +
                    " bytes of response data for extended frame. Got " +
                    responseFrame.size,
                )
                return null
            }
            val dataLengthM = responseFrame[5]
            val dataLengthL = responseFrame[6]
            val lengthChecksum = responseFrame[7]
            expectedDataLength = (dataLengthM.toUInt() shl 8) + dataLengthL.toUInt()

            if ((lengthChecksum + dataLengthM + dataLengthL) and 0xFF != 0) {
                Log.e(
                TAG,
                "Extended Frame failed length checksum. lengthChecksum: " +
                    lengthChecksum +
                    ", dataLengthM: " +
                    dataLengthM +
                    ", dataLengthL: " +
                    dataLengthL +
                    ", responseFrame: " +
                    responseFrame.toHex() +
                    ", data: " +
                    data,
                )
            }
        } else {
            if (responseFrame.size < NORMAL_FRAME_MIN_SIZE) {
                Log.e(
                TAG,
                "Expected at least " +
                    NORMAL_FRAME_MIN_SIZE +
                    " bytes of response data for normal frame. Got " +
                    responseFrame.size,
                )
                return null
            }

            val dataLength = responseFrame[3]
            val lengthChecksum = responseFrame[4]
            expectedDataLength = dataLength.toUByte().toUInt()

            if ((lengthChecksum + dataLength) and 0xFF != 0) {
                Log.e(
                TAG,
                "Normal Frame failed length checksum. lengthChecksum: " +
                    lengthChecksum +
                    ", dataLength: " +
                    dataLength +
                    ", responseFrame: " +
                    responseFrame.toHex() +
                    ", data: " +
                    data,
                )
            }
        }

        val tfi = if (isExtendedFrame) responseFrame[8] else responseFrame[5]
        if (tfi != 0xD5.toByte()) {
            Log.e(TAG, "Unexpected TFI Byte: Got " + tfi + ", expected 0xD5")
        }

        val dataPacket: ByteArray?
        val dataCheckSum: Byte
        val postAmble: Byte
        if (secondRead) {
            dataPacket =
                if (isExtendedFrame) {
                responseFrame.slice(9..responseFrame.size - 3).toByteArray()
                } else {
                responseFrame.slice(6..responseFrame.size - 3).toByteArray()
                }
            dataCheckSum = responseFrame[responseFrame.size - 2]
            postAmble = responseFrame[responseFrame.size - 1]
        } else {
            dataPacket =
                if (isExtendedFrame) {
                data.slice(15..data.size - 3).toByteArray()
                } else {
                data.slice(12..data.size - 3).toByteArray()
                }
            dataCheckSum = data[data.size - 2]
            postAmble = data[data.size - 1]
        }

        if (dataPacket.size != 0 && dataPacket.size.toUInt() != expectedDataLength - 1.toUInt()) {
            Log.e(
                TAG,
                "Unexpected data packet size: Got " +
                        dataPacket.size.toUInt() +
                        ", expected " +
                        (expectedDataLength - 1.toUInt()).toString() +
                        ",",
            )
        }

        val sum = dataPacket.sum()

        if ((tfi + sum + dataCheckSum) and 0xFF != 0) {
            Log.e(
                TAG,
                "Frame failed data checksum. TFI: " +
                        tfi +
                        ", sum: " +
                        sum +
                        ", secondFrame: " +
                        responseFrame.toHex(),
            )
        }

        if (postAmble != 0x00.toByte()) {
            if (tfi != 0xD5.toByte()) {
                Log.e(TAG, "Unexpected postamble byte when performing read - got " + responseFrame[4])
            }
            return null
        }

        transportLayer.write(ACK)
        Log.d(TAG, "Received frame - " + responseFrame.toHex() + ", dataPacket: " + dataPacket.toHex())

        return dataPacket
    }

    fun ByteArray.sum(): Byte {
        var sum = 0
        for (byte in this) {
            sum += byte
        }
        return sum.toByte()
    }

    private fun crc16a(data: ByteArray): ByteArray {
        var w_crc = 0x6363
        for (byte in data) {
            var newByte = byte.toInt() xor (w_crc and 0xFF)
            newByte = (newByte xor newByte shl 4) and 0xFF
            w_crc = ((w_crc shr 8) xor (newByte shl 8) xor (newByte shl 3) xor (newByte shr 4)) and 0xFF
        }

        return byteArrayOf((w_crc and 0xFF).toByte(), ((w_crc shr 8) and 0xFF).toByte())
    }

    private fun withCrc16a(data: ByteArray): ByteArray {
        return data + crc16a(data)
    }

    private fun constructFrame(data: ByteArray): ByteArray {
        val isExtendedFrame = data.size > NORMAL_FRAME_DATA_SIZE
        var frame: ByteArray =
            if (isExtendedFrame) {
                byteArrayOf(
                0x00,
                0x00,
                0xFF.toByte(),
                0xFF.toByte(),
                0xFF.toByte(),
                (data.size / 256).toByte(),
                (data.size % 256 + 1).toByte(),
                ((data.size / 256 + data.size % 256 + 1 and 0xFF).inv() + 0x01).toByte(),
                0xD4.toByte(),
                )
            } else {
                byteArrayOf(
                0x00,
                0x00,
                0xFF.toByte(),
                (data.size + 1).toByte(),
                ((data.size + 1 and 0xFF).inv() + 0x01).toByte(),
                0xD4.toByte(),
                )
            }
        var sum = 0xD4
        for (byte in data) {
            sum += byte
        }
        frame += (data)
        frame += ((sum.inv() and 0xFF) + 0x01).toByte()
        frame += (0x00).toByte()
        Log.d(TAG, "constructFrame::isExtendedFrame: " + isExtendedFrame + " Frame: " + frame.toHex())

        return frame
    }

    private fun ByteArray.toHex(): String =
        joinToString(separator = "") { eachByte -> "%02x".format(eachByte) }

    private fun String.decodeHex(): ByteArray {
        check(length % 2 == 0) { "Must have an even length" }
        return chunked(2).map { it.toInt(16).toByte() }.toByteArray()
    }

    companion object {
        private const val TAG = "PN532"
        val ACK: ByteArray =
            byteArrayOf(
                0x00.toByte(),
                0x00.toByte(),
                0xFF.toByte(),
                0x00.toByte(),
                0xFF.toByte(),
                0x00.toByte(),
            )
        private const val SAM_CONFIGURATION = 0x14.toByte()
        private const val IN_LIST_PASSIVE_TARGET = 0x4A.toByte()
        private const val RF_CONFIGURATION = 0x32.toByte()
        private const val WRITE_REGISTER = 0X08.toByte()
        private const val IN_COMMUNICATE_THRU = 0x42.toByte()
        private const val IN_DATA_EXCHANGE = 0x40.toByte()
        private val LONG_PREAMBLE = ByteArray(20)
        private const val MAX_READ_BUFFER_SIZE = 288
        private const val NORMAL_FRAME_MIN_SIZE = 8
        private const val NORMAL_FRAME_DATA_SIZE = 255
        private const val EXTENDED_FRAME_MIN_SIZE = 11
        private const val ACK_FRAME_SIZE = 6
    }
}
